src/add-text-track-data.js

```diff
@@ -66,19 +66,38 @@ const addTextTrackData = function(sourceHandler, captionArray, metadataArray) {
   if (metadataArray) {
     metadataArray.forEach(function(metadata) {
       let time = metadata.cueTime + this.timestampOffset;
+      let videoDuration = Number.isFinite(this.mediaSource_.duration) ?
+        this.mediaSource_.duration :
+        Number.MAX_VALUE;
 
       metadata.frames.forEach(function(frame) {
         let cue = new Cue(
-            time,
-            time,
-            frame.value || frame.url || frame.data || '');
+          time,
+          videoDuration,
+          frame.value || frame.url || frame.data || '');
 
         cue.frame = frame;
         cue.value = frame;
         deprecateOldCue(cue);
+
         this.metadataTrack_.addCue(cue);
       }, this);
     }, sourceHandler);
+
+    /** Updating the metadeta cues so that
+     * the endTime of each cue is the startTime of the next cue
+     * the endTime of last cue is the duration of the video
+     */
+    if (sourceHandler.metadataTrack_ && sourceHandler.metadataTrack_.cues) {
+      for (let i = 0; i < sourceHandler.metadataTrack_.cues.length - 1; i++) {
```

Comment:
My only reservation about this is that we will be iterating over all the cues every segment append. There may not ever be a large number of cues but this algorithm has a complexity of O(n^2) which has scary implications in extreme cases.

TBH, I'm not sure there is a good way to make this take less time without vastly complicating the algorithm but I felt like pointing this out since that kind of complexity can cause performance issues if you aren't careful.